### PR TITLE
rustc: Update LLVM

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit updates the LLVM submodule in use to the current HEAD of the LLVM
repository. This is primarily being done to start picking up unwinding support
for MSVC, which is currently unimplemented in the revision of LLVM we are using.
Along the way a few changes had to be made:
- As usual, lots of C++ debuginfo bindings in LLVM changed, so there were some
  significant changes to our RustWrapper.cpp
- As usual, some pass management changed in LLVM, so clang was re-scrutinized to
  ensure that we're doing the same thing as clang.
- Some optimization options are now passed directly into the
  `PassManagerBuilder` instead of through CLI switches to LLVM.
- The `NoFramePointerElim` option was removed from LLVM, favoring instead the
  `no-frame-pointer-elim` function attribute instead.
- The `LoopVectorize` option of the LLVM optimization passes has been disabled
  as it causes a divide-by-zero exception to happen in LLVM for zero-sized
  types. This is reported as https://llvm.org/bugs/show_bug.cgi?id=23763

Additionally, LLVM has picked up some new optimizations which required fixing an
existing soundness hole in the IR we generate. It appears that the current LLVM
we use does not expose this hole. When an enum is moved, the previous slot in
memory is overwritten with a bit pattern corresponding to "dropped". When the
drop glue for this slot is run, however, the switch on the discriminant can
often start executing the `unreachable` block of the switch due to the
discriminant now being outside the normal range. This was patched over locally
for now by having the `unreachable` block just change to a `ret void`.

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @brson 

cc @rust-lang/compiler - something to be aware of at least, and I may have updated something horribly incorrectly

**Comment by pnkfelix** [comment]
@eddyb notes on [a comment from #23535](https://github.com/rust-lang/rust/pull/23535#issuecomment-109266575) that we may want to restrict the `ret void` change so that it is only injected on the switch injected in the generated drop-glue.

**Comment by sanxiyn** [comment]
Apparently, LLVM PR23763 is fixed now.

**Comment by alexcrichton** [comment]
@pnkfelix I think we're covered on that front because the function `iter_structural_ty` is only called from generating the drop glue.

@sanxiyn it has indeed! I'm testing out pulling in that revision and running tests now

**Comment by eddyb** [comment]
@alexcrichton oh, of course, I forgot that. Maybe that machinery should be moved out of `trans::base` then (in the future).

**Comment by alexcrichton** [comment]
Looks like https://github.com/rust-lang/llvm/commit/47dfcb7745766fee1aaa135a212b3e1f39d81e96 did indeed fix the problems I was seeing! Updated the LLVM submodule.

File: src/rustllvm/PassWrapper.cpp [hunk 1]
```diff
@@ -123,12 +127,31 @@ LLVMRustAddAnalysisPasses(LLVMTargetMachineRef TM,
...
+                                    bool MergeFunctions,
+                                    bool SLPVectorize,
+                                    bool LoopVectorize) {
+#if LLVM_VERSION_MINOR >= 7
+    unwrap(PMB)->MergeFunctions = MergeFunctions;
+    unwrap(PMB)->SLPVectorize = SLPVectorize;
+    unwrap(PMB)->OptLevel = OptLevel;
+    unwrap(PMB)->LoopVectorize = LoopVectorize;
+#endif
 }
```
Comment (brson) [inline src/rustllvm/PassWrapper.cpp]: Now that the PMB opt level is set here and not through `llvm::LLVMPassManagerBuilderSetOptLevel` it appears that for LLVM < 7 we don't set the PMB opt level. Is that true, and is that bad?

File: src/rustllvm/RustWrapper.cpp [hunk 2]
```diff
@@ -684,11 +746,15 @@ extern "C" LLVMValueRef LLVMDIBuilderCreateDebugLocation(
...
+                                       unwrapDI<MDNode>(InlinedAt));
 
 #if LLVM_VERSION_MINOR >= 6
-    return wrap(MetadataAsValue::get(context, debug_loc.getAsMDNode(context)));
+    return wrap(MetadataAsValue::get(context, debug_loc.getAsMDNode(
+#if LLVM_VERSION_MINOR <= 6
+            context
+#endif
+        )));
 #else
```
Comment (brson) [inline src/rustllvm/RustWrapper.cpp]: Yikes... that's lovely.

File: src/test/codegen/loads.rs [hunk 3]
```diff
@@ -46,7 +46,7 @@ pub fn small_array_alignment(x: [i8; 4]) -> [i8; 4] {
 // dependent alignment
 #[no_mangle]
 pub fn small_struct_alignment(x: Bytes) -> Bytes {
-// CHECK: [[VAR:%[0-9]+]] = load i32* %{{.*}}, align 1
+// CHECK: [[VAR:%[0-9]+]] = load i32, i32* %{{.*}}, align 1
 // CHECK: ret i32 [[VAR]]
     x
```
Comment (brson) [inline src/test/codegen/loads.rs]: What are these test changes about?

**Comment by brson** [comment]
r=me once you've answered my q's.

File: src/test/codegen/loads.rs (hunk 3 above)
Comment (eddyb) [inline src/test/codegen/loads.rs]: LLVM is switching to opaque pointers, so instructions working with pointers have to supply both the pointer type and the pointee in the interim.
The final result will have only the pointee type, the pointer type always being `ptr` or something like that.

File: src/rustllvm/PassWrapper.cpp (hunk 1 above)
Comment (alexcrichton) [inline src/rustllvm/PassWrapper.cpp]: Oh this reminded me that I meant to recompile these files against 3.5 and 3.6, and now having done so it looks like only `MergeFunctions` wasn't available in 3.5 but everything else is now set on all versions.

File: src/rustllvm/RustWrapper.cpp (hunk 2 above)
Comment (alexcrichton) [inline src/rustllvm/RustWrapper.cpp]: Yeah this file is... a mess. Supporting multiple LLVM versions is definitely not our forte.

File: src/test/codegen/loads.rs (hunk 3 above)
Comment (alexcrichton) [inline src/test/codegen/loads.rs]: Yeah this was just changes syntactically in the IR self, semantically nothing changed here.

**Comment by alexcrichton** [comment]
This is now blocked on https://llvm.org/bugs/show_bug.cgi?id=23779

**Comment by alexcrichton** [comment]
Ok, the LLVM bug has not been fixed yet, and I had some time to do some more investigation. The end result was adding [this extra commit](https://github.com/rust-lang/llvm/commit/993e1fe0f6520dfbc9bc417d2a6ae62ea177df58) to the LLVM branch that we're using. Still running tests locally, but I suspect it will fix *that* segfault at least.

re-r? @brson

**Comment by brson** [comment]
r=me with comment addressed

**Comment by brson** [comment]
Huh, I don't know where my comment went, but please add a comment to the source explaining the GOT issue so someone might be clued in to the need to revert it later.

**Comment by alexcrichton** [comment]
*crosses fingers*

**Comment by retep998** [comment]
Do you think you could fix this error before trying again?
```
In file included from /Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/src/llvm/lib/ExecutionEngine/Orc/OrcMCJITReplacement.cpp:10:
In file included from /Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/src/llvm/lib/ExecutionEngine/Orc/OrcMCJITReplacement.h:20:
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/src/llvm/include/llvm/ExecutionEngine/Orc/LazyEmittingLayer.h:126:66: error: only virtual member functions can be marked 'override'
                                 bool ExportedSymbolsOnly) const override {
                                                                 ^~~~~~~~
1 error generated.
make[4]: *** [/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-nopt-t/build/obj/x86_64-apple-darwin/llvm/lib/ExecutionEngine/Orc/Release/OrcMCJITReplacement.o] Error 1
make[3]: *** [Orc/.makeall] Error 2
make[2]: *** [ExecutionEngine/.makeall] Error 2
make[2]: *** Waiting for unfinished jobs....
```

**Comment by KindDragon** [comment]
Congratulations! :+1: :smile:

**Comment by frewsxcv** [comment]
This caused a regression related to LLVM passes. I fixed it in https://github.com/rust-lang/rust/pull/31176

**Pull request merged**